caffe2/operators/expand_op.h

```diff
@@ -0,0 +1,149 @@
+#ifndef CAFFE2_OPERATORS_REDUCE_OPS_H_
+#define CAFFE2_OPERATORS_REDUCE_OPS_H_
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+
+#include "caffe2/core/context.h"
+#include "caffe2/core/operator.h"
+#include "caffe2/core/types.h"
+#include "caffe2/utils/math.h"
+
+namespace caffe2 {
+
+template <typename InputTypes, class Context, class Expander>
+class ExpandOp final : public Operator<Context> {
+ public:
+  USE_OPERATOR_CONTEXT_FUNCTIONS;
+
+  ExpandOp(const OperatorDef& operator_def, Workspace* ws)
+      : Operator<Context>(operator_def, ws) {}
+
+  bool RunOnDevice() override {
+    return DispatchHelper<InputTypes>::call(this, Input(0));
+  }
+ template <typename T>
+  bool DoRunWithType() {
+    const auto& X = Input(0);
+    const auto& Y_shape_tensor = Input(1);
+	std::vector<int> shape_dims(Y_shape_tensor.size());
+	context_.template Copy<int, Context, CPUContext>(Y_shape_tensor.size(), Y_shape_tensor.template data<int>(), shape_dims.data());
+    const int* Y_shape = Y_shape_tensor.template data<int>();
+    auto* Y = Output(0);
+
+	const int ndim = shape_dims.size();
+    const std::vector<int> X_dims(X.dims().cbegin(), X.dims().cend());
+    std::vector<int> Y_dims;
+    Y_dims.reserve(std::max(ndim, X.ndim()));
+    // ndim, X.ndim() might equal to 0
+    std::cout << "loginfoo " << ndim << " " << X.ndim() << std::endl;
+    for (int i = ndim - 1, j = X.ndim() - 1; i >= 0 || j >= 0; i--, j--) {
+      const int shape_x = (j >= 0 ? X_dims[j] : 1);
+      const int shape_y = (i >= 0 ? shape_dims[i] : 1);
+      CAFFE_ENFORCE(
+          shape_x == 1 || shape_y == 1 || shape_x == shape_y,
+          "Dimensions format invalid.");
+      Y_dims.push_back(std::max(shape_x, shape_y));
+    }
+    std::reverse(Y_dims.begin(), Y_dims.end());
+    Y->Resize(Y_dims);
+    return expander_.template Forward<T>(
+        X.ndim(),
+        X_dims,
+        Y_dims.size(),
+        Y_dims,
+        X.template data<T>(),
+        Y->template mutable_data<T>(),
+        &context_);
+  }
+
+ private:
+  Expander expander_{};
```

Comment:
If we don't have any plan for some other expand ops. I this there is no need for the Expander part. It is kind of over design. I'd suggest make the logic of Forward and Backward   in RunOnDevice in ExpandOp and ExpandGradientOp directly.